File: Week 2/BOJ_실버 2_랜선 자르기/hyungjoon_solved.py
'''
문제 : 랜선 자르기
난이도 : 실버 2
소요 시간 : 20분
'''
import sys
se = sys.stdin.readline
answer = 0

k, n = map(int, se().split())
arr = [ int(se()) for _ in range(k)]

# 이분탐색으로 진행
s, e = 1, max(arr)

while s <= e:
    mid = (s+e) // 2
    lines = 0
    for i in arr:
        lines += i//mid
    
    # 자른 랜선이 목표치에 도달하지 못하면, 더 짧게 잘라야 하니 e = mid - 1
    if lines < n:
        e = mid - 1
    # 목표치를 넘으면 더 길게 잘라야 하니 s = mid + 1
    else:
        s = mid + 1

print(e)

File: Week 2/BOJ_실버 1_겹치는 건 싫어/hyungjoon_solved.java
package backjoon.Binary_Search.Silver_1_20922;

import java.util.*;
import java.io.*;

/**
 * 문제 : 겹치는 건 싫어
 * 난이도 : 실버 1
 * 소요 시간 : 40분
 */
public class Main {
    public static void main(String[] args) throws IOException{
        // 입력
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        int[] arr = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();

        // 숫자별 갯수를 cnt 하기 위한 배열
        int[] cnt = new int[200001];
        int answer = 0;

        // l, r 모두 0부터 시작하는 투포인터 사용
        int l = 0, r = 0;
        while(r < n) {
            // r을 늘리면서 해당하는 수의 개수가 k개 이하일 때 까지 쭉쭉 늘려준다.
            while (r < n && cnt[arr[r]] + 1 <= k) {
                cnt[arr[r]]++;
                r++;
            }
            // 길이 계산
            int temp = r - l;
            answer = Math.max(temp, answer);
            // 왼쪽 포인터 당겨주고 수의 개수를 차감시켜준다.
            cnt[arr[l]]--;
            l++;
        }
        System.out.println(answer);
    }
}
